### PR TITLE
Add doc build

### PR DESCRIPTION
**Pull request by svekars**
This PR adds:
- doc/ dir with the main doc structure and Sphinx config
- doc_build.yml - a github action that builds the documentation.
- entries in the .gitignore files to ignore local build artifacts.
- Landing page

Local build screenshot:
Screenshot 2024-04-15 at 3 01 20 PM
Screenshot 2024-04-15 at 3 01 33 PM
Screenshot 2024-04-15 at 3 01 41 PM
Screenshot 2024-04-15 at 3 02 55 PM

### CODE REVIEW

**Comment by msaroufim** [comment]
Nice! Lmk when you need a review!

File: docs/source/api_ref_dtypes.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+.. autosummary::
+    :toctree: generated/
+    :nosignatures:
+
+    to_nf4
+    UInt4Tensor
+
+..
+  _NF4Tensor - add after fixing torchao/dtypes/nf4tensor.py:docstring of torchao.dtypes.nf4tensor.NF4Tensor.dequantize_scalers:6:Unexpec   ted indentation.
```
Comment (jerryzh168) [inline docs/source/api_ref_dtypes.rst:17]: nit: seems there are some extra space after `Unexpec`

**Review by jerryzh168 — APPROVED** [review]

File: docs/source/api_ref_quantization.rst [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    dynamically_quantize_per_tensor
+    quantize_activation_per_token_absmax
+    dynamically_quantize_per_channel
+    dequantize_per_tensor
+    dequantize_per_channel
+    quant_int8_dynamic_linear
+    quant_int8_matmul
+    quant_int8_dynamic_per_token_linear
+    quant_int8_per_token_matmul
+    get_scale
```
Comment (jerryzh168) [inline docs/source/api_ref_quantization.rst:29]: can you remove these

File: docs/source/api_ref_quantization.rst [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    SmoothFakeDynamicallyQuantizedLinear
+    swap_linear_with_smooth_fq_linear
+    smooth_fq_linear_to_inference
+    set_smooth_fq_attribute
+    Int8DynamicallyQuantizedLinearWeight
+    Int8WeightOnlyQuantizedLinearWeight
+    Int4WeightOnlyQuantizedLinearWeight
+    compute_error
+    get_model_size_in_bytes
+    WeightOnlyInt8QuantLinear
```
Comment (jerryzh168) [inline docs/source/api_ref_quantization.rst:40]: and these

File: docs/Makefile [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+#
+
+ifneq ($(EXAMPLES_PATTERN),)
+    EXAMPLES_PATTERN_OPTS := -D sphinx_gallery_conf.filename_pattern="$(EXAMPLES_PATTERN)"
+endif
+
+# You can set these variables from the command line.
+SPHINXOPTS    = -W -j auto $(EXAMPLES_PATTERN_OPTS)
+SPHINXBUILD   = sphinx-build
+SPHINXPROJ    = torchtune
```
Comment (msaroufim) [inline docs/Makefile:11]: i still see 2 places with torchtune mentioned, doesnt seem like it breaks anything but might as well fix

**Review by msaroufim — APPROVED** [review]

**Pull request merged**